### PR TITLE
Introduce support types logic

### PR DESCRIPTION
**Pull request by ShockedPlot7560**
## Introduction
Many block cannot be placed on certain types of blocks. In addition, there is currently no way for blocks to specify the type of their surface for block placement

## The concept 
It all comes down to the new feature that has been added: ``getSupportType(Facing) : SupportType``. It will then return 4 different types of surfaces according to this table: 
Type | Description
----- | -----
Full | The support face doesn't have any restrictions on support it can provide.
Edge | Support any element which reach the eadges of the block : repeaters,comparator on top of hoppers, cauldron)
Center | Support any elements which can placed in the center (torch on top of fence or wall)
None | No placement is allowed

The discussion and further explanation of this concept can be found here : https://github.com/pmmp/PocketMine-MP/issues/4856

### Relevant issues
* #4856 
* https://github.com/pmmp/PocketMine-MP/issues/458
* https://github.com/pmmp/PocketMine-MP/issues/4529
* https://github.com/pmmp/PocketMine-MP/issues/3299

Some blocks have also had their placement corrected such as ``RedstoneWire`` which was not checking any support

### API changes
* Add ``getSupportType()`` in ``Block``
* Add ``SupportType`` classes

### Behavioural changes
Any block that inherits from the parent class : ``Block`` will have a ``None`` surface by default. The default value for ``Opaque`` is: ``Full`` 

## Backwards compatibility
Some block placements have been reviewed and fixed but some may have been impacted due to the change in behavior of the block placement

## Tests
Take the blocks that have been modified and test their placement one by one in all possible positions and situations. 
These were done during development and would be too long to reproduce on video.

### CODE REVIEW

File: src/block/Stair.php [hunk 1]
```diff
@@ -128,4 +129,11 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
 
 		return parent::place($tx, $item, $blockReplace, $blockClicked, $face, $clickVector, $player);
 	}
+
+	public function getSupportType(int $facing) : SupportType{
+		if($facing === Facing::UP && $this->isUpsideDown() || $facing === $this->facing || $facing === Facing::DOWN && !$this->isUpsideDown()){
+			return SupportType::FULL();
```
Comment (dktapps) [inline src/block/Stair.php]: This needs to include logic for corner stairs.

Comment (dktapps) [reply src/block/Stair.php]: This is also potentially subject to precedence issues (needs more ().)

Comment (ShockedPlot7560) [reply src/block/Stair.php]: in fact, the corner stairs are in this case problematic

Comment (ShockedPlot7560) [reply src/block/Stair.php]: Do we really implement as vanilla or do we add this feature: 
```
Take a corner stair type of INNER_RIGHT facing north.
In vanilla, you will be able to put torch on the north face and not on the east.
Knowing that east face are full, should we allowed the torch placement or not ?
```

**Comment by dktapps** [comment]
Hi, I haven't had time to review this yet. It's a big PR and I've been busy with other things the last couple of weeks. I'll get to it soon.

**Comment by ShockedPlot7560** [comment]
Ok no problem.

**Comment by dktapps** [comment]
A couple of suggestions come to mind (I haven't reviewed the whole thing yet):

`SupportType->hasEdgeSupport()` and `SupportType->hasCenterSupport()` would probably be useful, because:

| Support type | Supports edges | Supports center |
|:--------------:|:------------------:|:------------------:|
| None | | |
| Center | | YES |
| Edge | YES | |
| Full | YES | YES |

Typically this means that we're usually going to be checking for two types of support: `FULL` + one of `CENTER` or `EDGE`.

File: src/block/Block.php [hunk 2]
```diff
@@ -642,4 +643,8 @@ public function calculateIntercept(Vector3 $pos1, Vector3 $pos2) : ?RayTraceResu
 
 		return $currentHit;
 	}
+
+	public function getSupportType(int $facing) : SupportType{
+		return SupportType::NONE();
```
Comment (dktapps) [inline src/block/Block.php]: I'm not sure this really makes sense. `Block` assumes the presence of a 1x1x1 collision cube, which normally equates to full support.

**Review by dktapps — COMMENTED** [review]
Concept is OK, although there's definitely ways to make the code clearer, and some of the choices for support types are a bit weird. Needs more work.

**Comment by dktapps** [comment]
Things noticed so far:
- Hopper's upper face seems to be FULL, rather than EDGE. This is counterintuitive, but it's the only way I can explain the behaviour I see (torch can be placed on top of hopper, but not on top of stuff like cauldron or composter, while rails can)
- Spawner is weird. It accepts placement of torches, but not stuff like buttons, levers, pressure plates etc. I don't know what to make of it.
- Buttons should require CENTER support type
- Flower pot should require CENTER support type
- Lanterns should require CENTER support type
- Lever should require CENTER support type
- Torch requires CENTER support when placed on the top face of a block. It can't be placed on top of a cauldron or composter.

... will update via edits while testing

File: src/block/Block.php (hunk 2 above)
Comment (ShockedPlot7560) [reply src/block/Block.php]: Indeed, this part should be as ``Full``, and the ``Transparent`` should be modified for a none support

File: src/block/DeadBush.php [hunk 3]
```diff
@@ -34,15 +35,15 @@
 class DeadBush extends Flowable{
 
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
-		if(!$this->getSide(Facing::DOWN)->isTransparent()){
+		if($this->getSide(Facing::DOWN)->getSupportType(Facing::UP)->equals(SupportType::FULL())){
```
Comment (dktapps) [inline src/block/DeadBush.php]: I don't think this needs to be changed

File: src/block/Lantern.php [hunk 4]
```diff
@@ -77,7 +78,10 @@ protected function canAttachTo(Block $b) : bool{
 	}
 
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
-		if(!$this->canAttachTo($this->position->getWorld()->getBlock($blockReplace->getPosition()->up())) && !$this->canAttachTo($this->position->getWorld()->getBlock($blockReplace->getPosition()->down()))){
+		if($face === Facing::DOWN && $blockReplace->getSide(Facing::UP)->getSupportType(Facing::DOWN)->equals(SupportType::NONE())){
+			return false;
+		}
+		if($face === Facing::UP && $blockReplace->getSide(Facing::DOWN)->getSupportType(Facing::UP)->equals(SupportType::NONE())){
```
Comment (dktapps) [inline src/block/Lantern.php]: This can be simplified to check for `Axis::Y` and use `Facing::opposite()` to get the appropriate facings.

File: src/block/Hopper.php [hunk 5]
```diff
@@ -99,5 +100,12 @@ public function onScheduledUpdate() : void{
 		//TODO
 	}
 
+	public function getSupportType(int $facing) : SupportType{
+		if($facing === Facing::UP){
+			return SupportType::EDGE();
```
Comment (dktapps) [inline src/block/Hopper.php]: It's unclear what the correct behaviour of this should be. Many blocks (e.g. torches, lanterns) can be placed on top of a hopper, but can't be placed on top of other EDGE blocks (such as composters); however, rails can be placed on both.

I think the correct thing for hopper would be to treat the top face as FULL. I don't like it, but it seems to be how the game itself treats it.

File: src/block/RedMushroom.php [hunk 6]
```diff
@@ -43,7 +44,7 @@ public function onNearbyBlockChange() : void{
 
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
 		$down = $this->getSide(Facing::DOWN);
-		if(!$down->isTransparent()){
+		if(!$down->getSupportType(Facing::UP)->equals(SupportType::FULL())){
```
Comment (dktapps) [inline src/block/RedMushroom.php]: I don't think this should be changed

File: src/block/Vine.php [hunk 7]
```diff
@@ -116,7 +117,7 @@ protected function recalculateCollisionBoxes() : array{
 	}
 
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
-		if(!$blockClicked->isSolid() || Facing::axis($face) === Axis::Y){
+		if(!$blockClicked->getSupportType($face)->equals(SupportType::FULL()) || Facing::axis($face) === Axis::Y){
```
Comment (dktapps) [inline src/block/Vine.php]: I don't think this is necessary

**Comment by dktapps** [comment]
The following blocks need to be changed:
- Doors (any FULL/EDGE block is sufficient; they should be placeable on slabs, composters and cauldrons)
- Beds (CENTER/EDGE/FULL all seem to be OK)
- Dead bush (these require certain specific blocks, not support types)
- Lantern (the placement logic has not been completely updated)
- Mushroom (the logic has been changed incorrectly; the original logic is correct aside from the missing lightlevel check)

**Comment by dktapps** [comment]
- Bell (any support seems to be OK; they can be placed on full blocks, composters, and fences, but not on beds)
- Snow layer (appears to require CENTER support, as best I can tell)
- Coral (appears to require CENTER support)
- Coral fan (appears to require CENTER support)

**Comment by ShockedPlot7560** [comment]
> Spawner is weird. It accepts placement of torches, but not stuff like buttons, levers, pressure plates etc. I don't know what to make of it.

Looking at the wiki and practicing my vanilla tests, we can see that the spawner accepts torches but not other blocks. This can be explained by the fact that it is transparent, and thus everything that is lever, button, pressure plate is useless (does not transmit any current)

**Comment by dktapps** [comment]
That's not enough to explain it. Glass has all those same properties, and yet still accepts the placement of all of these blocks.

**Comment by ShockedPlot7560** [comment]
It is true, in this case, it is necessary to leave in a logic external to minecraft vanilla and allow the placement of all

**Comment by ShockedPlot7560** [comment]
> Lantern (the placement logic has not been completely updated)

Are you talking about the check to define if the lantern is hanging?
```php
$this->hanging = ($face === Facing::DOWN || !$this->canAttachTo($this->position->getWorld()->getBlock($blockReplace->getPosition()->down())));
```

**Comment by ShockedPlot7560** [comment]
All requested changes have been made. The only things pending, waiting for an answer, for debate are: 
- Spawner: Should we give them a ``FULL`` support for all surfaces, or should we try to stick to minecraft vanilla as currently.
- Lantern: Which investment is currently unimplemented, missing?

File: src/block/Transparent.php [hunk 8]
```diff
@@ -23,9 +23,15 @@
...
+use pocketmine\block\utils\SupportType;
+
 class Transparent extends Block{
 
 	public function isTransparent() : bool{
 		return true;
 	}
+
+	public function getSupportType(int $facing) : SupportType{
+		return SupportType::NONE();
```
Comment (dktapps) [inline src/block/Transparent.php]: This is incorrect. Many transparent blocks are still full cubes.

**Comment by dktapps** [comment]
> Spawner: Should we give them a FULL support for all surfaces, or should we try to stick to minecraft vanilla as currently.

Spawner shouldn't have any support, since that's how it appears to behave.

File: src/block/RedMushroom.php [hunk 9]
```diff
@@ -42,11 +41,18 @@ public function onNearbyBlockChange() : void{
...
 
 		return false;
 	}
+
+	protected function canBeSupportedBy(Block $block) : bool{
+		$position = $block->getPosition();
+		$lightLevel = $position->getWorld()->getFullLightAt($position->x, $position->y, $position->z);
+		$id = $block->getId();
+		// TODO: nylium support
+		return ($lightLevel <= 12 && !$block->isTransparent()) || $id === BlockLegacyIds::MYCELIUM || $id === BlockLegacyIds::PODZOL;
```
Comment (dktapps) [inline src/block/RedMushroom.php]: The light level here will always be zero if the block is opaque.

**Comment by dktapps** [comment]
> Lantern: Which investment is currently unimplemented, missing?

You've left a usage of `canAttachTo()` which has inconsistent behaviour with the new code.

File: src/block/Transparent.php (hunk 8 above)
Comment (ShockedPlot7560) [reply src/block/Transparent.php]: I agree on this point, however, some blocks are not real cube and do not require special support that ``NONE``.  
Instead of defining the supports of the transparent blocks one by one in ``NONE``, it is better to do the case by case of transparent blocks not having a null support ?

Comment (dktapps) [reply src/block/Transparent.php]: No, I don't agree. For `Flowable` this would make sense, but not for `Transparent`.

File: src/block/Lantern.php [hunk 10]
```diff
@@ -81,7 +81,7 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
 			return false;
 		}
 
-		$this->hanging = ($face === Facing::DOWN || !$this->canAttachTo($this->position->getWorld()->getBlock($blockReplace->getPosition()->down())));
+		$this->hanging = ($face === Facing::DOWN || !$this->canBeSupportedBy($this->position->getWorld()->getBlock($blockReplace->getPosition()->down()), Facing::UP));
```
Comment (dktapps) [inline src/block/Lantern.php:85]: canAttachTo() is now dead code

File: src/block/MonsterSpawner.php [hunk 11]
```diff
@@ -42,11 +40,4 @@ protected function getXpDropAmount() : int{
 	public function onScheduledUpdate() : void{
 		//TODO
 	}
-
-	public function getSupportType(int $facing) : SupportType{
-		if($facing === Facing::UP || $facing === Facing::DOWN){
-			return SupportType::CENTER();
-		}
-		return SupportType::FULL();
-	}
```
Comment (dktapps) [inline src/block/MonsterSpawner.php]: This should have NONE support type

Comment (ShockedPlot7560) [reply src/block/MonsterSpawner.php]: Indeed, I thought ``Block`` was still setting ``NONE`` by default, so much for me

**Comment by ShockedPlot7560** [comment]
Is the PhpStan problem important? It's a typing problem, but casting the value as an integer doesn't seem like a good idea.

Looking at the uses of ``getFullLightAt``, I could notice that the use is the same, but no typing error came out.

**Comment by dktapps** [comment]
No, it's currently ignored via `actual-problems.neon` baseline

File: src/block/Fence.php [hunk 12]
```diff
@@ -93,4 +94,11 @@ protected function recalculateCollisionBoxes() : array{
 
 		return $bbs;
 	}
+
+	public function getSupportType(int $facing) : SupportType{
+		if(Facing::axis($facing) === Axis::Y){
+			return SupportType::CENTER();
+		}
+		return SupportType::NONE();
```
Comment (dktapps) [inline src/block/Fence.php]: This can be reduced to a ternary

File: src/block/Hopper.php [hunk 13]
```diff
@@ -99,5 +100,12 @@ public function onScheduledUpdate() : void{
 		//TODO
 	}
 
+	public function getSupportType(int $facing) : SupportType{
+		if($facing === Facing::UP){
+			return SupportType::FULL();
+		}
+		return SupportType::NONE();
+	}
```
Comment (dktapps) [inline src/block/Hopper.php]: This can be reduced to a ternary

File: src/block/SnowLayer.php [hunk 14]
```diff
@@ -81,7 +82,7 @@ protected function recalculateCollisionBoxes() : array{
 	}
 
 	private function canBeSupportedBy(Block $b) : bool{
-		return $b->isSolid() || ($b instanceof SnowLayer && $b->isSameType($this) && $b->layers === self::MAX_LAYERS);
+		return $b->getSupportType(Facing::UP)->hasCenterSupport() || ($b instanceof SnowLayer && $b->isSameType($this) && $b->layers === self::MAX_LAYERS);
```
Comment (dktapps) [inline src/block/SnowLayer.php]: I think the second condition can be removed with this change

File: src/block/Torch.php [hunk 15]
```diff
@@ -87,12 +89,21 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
...
+				if($this->canBeSupportBy($block, Facing::opposite($side))){
 					$this->facing = Facing::opposite($side);
 					return parent::place($tx, $item, $blockReplace, $blockClicked, $face, $clickVector, $player);
 				}
 			}
 		}
 		return false;
 	}
+
+	protected function canBeSupportBy(Block $support, int $face) : bool{
```
Comment (dktapps) [inline src/block/Torch.php]: Inconsistent function name

File: src/block/Wall.php [hunk 16]
```diff
@@ -76,4 +77,11 @@ protected function recalculateCollisionBoxes() : array{
 				->trim(Facing::EAST, $east ? 0 : $inset)
 		];
 	}
+
+	public function getSupportType(int $facing) : SupportType{
+		if($facing === Facing::UP || $facing === Facing::DOWN){
```
Comment (dktapps) [inline src/block/Wall.php]: Facing::axis can be used here

File: src/block/Wall.php [hunk 17]
```diff
@@ -76,4 +77,11 @@ protected function recalculateCollisionBoxes() : array{
...
+			return SupportType::CENTER();
+		}
+		return SupportType::NONE();
+	}
```
Comment (dktapps) [inline src/block/Wall.php:84]: This can be reduced to a ternary

**Comment by dktapps** [comment]
To clarify the reasoning for the inversion of default support types:

Support type differences generally go hand-in-hand with collision box shape differences (with few exceptions). Therefore, since the default behaviour is to assume that blocks have a full collision box, they should also be assumed to have full support on all faces unless otherwise specified. This is why Transparent can't declare NONE support.

I'm aware that this results in more changed files than it would otherwise have done, but this way makes more sense - you'll notice that usually the changes sit right next to `recalculateCollisionBoxes()` or something similar.

**Comment by ShockedPlot7560** [comment]
That was the behavior I understood in the end. It was not my logic at first, but now it seems more coherent with the rest

File: src/block/BaseRail.php [hunk 18]
```diff
@@ -220,11 +222,13 @@ private function setConnections(array $connections) : void{
...
 	public function onNearbyBlockChange() : void{
-		if($this->getSide(Facing::DOWN)->isTransparent()){
+		$supportType = $this->getSide(Facing::DOWN)->getSupportType(Facing::UP);
+		if($supportType->equals(SupportType::NONE()) || $supportType->equals(SupportType::CENTER())){
 			$this->position->getWorld()->useBreakOn($this->position);
 		}else{
 			foreach($this->getCurrentShapeConnections() as $connection){
-				if(($connection & RailConnectionInfo::FLAG_ASCEND) !== 0 && $this->getSide($connection & ~RailConnectionInfo::FLAG_ASCEND)->isTransparent()){
+				$supportType = $this->getSide($connection & ~RailConnectionInfo::FLAG_ASCEND)->getSupportType(Facing::UP);
+				if(($connection & RailConnectionInfo::FLAG_ASCEND) !== 0 && ($supportType->equals(SupportType::NONE()) || $supportType->equals(SupportType::CENTER()))){
```
Comment (dktapps) [inline src/block/BaseRail.php]: I'm not convinced this is correct. Slope support should depend on the face of the block it leans on, not the top face.

File: src/block/BaseRail.php [hunk 19]
```diff
@@ -38,7 +39,8 @@
 abstract class BaseRail extends Flowable{
 
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
-		if(!$blockReplace->getSide(Facing::DOWN)->isTransparent()){
+		$supportType = $blockReplace->getSide(Facing::DOWN)->getSupportType(Facing::UP);
+		if(!$supportType->equals(SupportType::NONE()) && !$supportType->equals(SupportType::CENTER())){
```
Comment (dktapps) [inline src/block/BaseRail.php]: `supportType->hasEdgeSupport()` can be used here

File: src/block/BaseRail.php [hunk 20]
```diff
@@ -220,11 +222,13 @@ private function setConnections(array $connections) : void{
 	}
 
 	public function onNearbyBlockChange() : void{
-		if($this->getSide(Facing::DOWN)->isTransparent()){
+		$supportType = $this->getSide(Facing::DOWN)->getSupportType(Facing::UP);
+		if($supportType->equals(SupportType::NONE()) || $supportType->equals(SupportType::CENTER())){
```
Comment (dktapps) [inline src/block/BaseRail.php]: `!supportType->hasEdgeSupport()` can be used here

File: src/block/BaseRail.php (hunk 18 above)
Comment (dktapps) [reply src/block/BaseRail.php]: `!supportType->hasEdgeSupport()` can be used here

File: src/block/Slab.php [hunk 21]
```diff
@@ -128,4 +129,13 @@ protected function recalculateCollisionBoxes() : array{
 	public function getDropsForCompatibleTool(Item $item) : array{
 		return [$this->asItem()->setCount($this->slabType->equals(SlabType::DOUBLE()) ? 2 : 1)];
 	}
+
+	public function getSupportType(int $facing) : SupportType{
+		if($this->getSlabType()->equals(SlabType::DOUBLE())){
```
Comment (dktapps) [inline src/block/Slab.php]: Getter usage unnecessary here

File: src/block/BaseRail.php (hunk 18 above)
Comment (ShockedPlot7560) [reply src/block/BaseRail.php]: It is not even either of the two logics. As can be tested in vanilla, by putting a hopper and rails, the rail seems to be based on the block below it all the time.

Comment (dktapps) [reply src/block/BaseRail.php]: Never mind, you are correct: 
image

File: src/block/BaseRail.php [hunk 22]
```diff
@@ -38,7 +38,8 @@
 abstract class BaseRail extends Flowable{
 
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
-		if(!$blockReplace->getSide(Facing::DOWN)->isTransparent()){
+		$supportType = $blockReplace->getSide(Facing::DOWN)->getSupportType(Facing::UP);
```
Comment (dktapps) [inline src/block/BaseRail.php]: Extra variable can be removed

File: src/block/BaseRail.php [hunk 23]
```diff
@@ -220,11 +221,13 @@ private function setConnections(array $connections) : void{
 	}
 
 	public function onNearbyBlockChange() : void{
-		if($this->getSide(Facing::DOWN)->isTransparent()){
+		$supportType = $this->getSide(Facing::DOWN)->getSupportType(Facing::UP);
```
Comment (dktapps) [inline src/block/BaseRail.php]: same here

File: src/block/BaseRail.php [hunk 24]
```diff
@@ -220,11 +221,13 @@ private function setConnections(array $connections) : void{
...
+		if(!$supportType->hasEdgeSupport()){
 			$this->position->getWorld()->useBreakOn($this->position);
 		}else{
 			foreach($this->getCurrentShapeConnections() as $connection){
-				if(($connection & RailConnectionInfo::FLAG_ASCEND) !== 0 && $this->getSide($connection & ~RailConnectionInfo::FLAG_ASCEND)->isTransparent()){
+				$supportType = $this->getSide($connection & ~RailConnectionInfo::FLAG_ASCEND)->getSupportType(Facing::UP);
```
Comment (dktapps) [inline src/block/BaseRail.php]: and here

File: src/block/FlowerPot.php [hunk 25]
```diff
@@ -104,19 +104,23 @@ protected function recalculateCollisionBoxes() : array{
 	}
 
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
-		if($this->getSide(Facing::DOWN)->isTransparent()){
+		if($this->canBeSupportedBy($this->getSide(Facing::DOWN), Facing::UP)){
```
Comment (dktapps) [inline src/block/FlowerPot.php]:
```suggestion
		if(!$this->canBeSupportedBy($this->getSide(Facing::DOWN), Facing::UP)){
```

File: src/block/Lantern.php [hunk 26]
```diff
@@ -72,22 +73,27 @@ protected function recalculateCollisionBoxes() : array{
 		];
 	}
 
-	protected function canAttachTo(Block $b) : bool{
-		return !$b->isTransparent();
-	}
-
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
-		if(!$this->canAttachTo($this->position->getWorld()->getBlock($blockReplace->getPosition()->up())) && !$this->canAttachTo($this->position->getWorld()->getBlock($blockReplace->getPosition()->down()))){
+		if(Facing::axis($face) === Axis::Y && !$this->canBeSupportedBy($blockReplace->getSide(Facing::opposite($face)), $face)){
```
Comment (dktapps) [inline src/block/Lantern.php]: This is incorrect. The block should look for up/down supports regardless of which face was clicked.

You can see the problem by right-clicking with a lantern on a wall when there are no available attachment points - the lantern item will be consumed, but no lantern is placed.

File: src/block/RedMushroom.php (hunk 9 above)
Comment (dktapps) [reply src/block/RedMushroom.php]: I'm not actually sure this logic belongs in this PR at all. It's unrelated to the objective, which is to introduce support types logic.

File: src/block/Sapling.php [hunk 27]
```diff
@@ -125,4 +124,12 @@ private function grow(?Player $player) : bool{
 	public function getFuelTime() : int{
 		return 100;
 	}
+
+	protected function hasValidSupport() : bool{
+		$down = $this->getSide(Facing::DOWN);
+		if($down->getId() === BlockLegacyIds::GRASS || $down->getId() === BlockLegacyIds::DIRT || $down->getId() === BlockLegacyIds::FARMLAND){
```
Comment (dktapps) [inline src/block/Sapling.php]: Not sure this change belongs in this PR.

File: src/block/Sugarcane.php [hunk 28]
```diff
@@ -118,11 +118,24 @@ public function onRandomTick() : void{
...
+	 * checks only the solid support of the sugar cane and not the proximity with water
+	 */
+	protected function canBeSupportedBy(Block $block) : bool{
+		$id = $block->getId();
+		//TODO: rooted dirt, moss block
+		return $id === BlockLegacyIds::GRASS
+			|| $id === BlockLegacyIds::DIRT
+			|| $id === BlockLegacyIds::PODZOL
+			|| $id === BlockLegacyIds::MYCELIUM
+			|| $id === BlockLegacyIds::SAND;
```
Comment (dktapps) [inline src/block/Sugarcane.php]: Not sure this change belongs in this PR.

File: src/block/RedMushroom.php (hunk 9 above)
Comment (ShockedPlot7560) [reply src/block/RedMushroom.php]: I'm going to move all this to other pull requests soon if not done before. (Light condition)

**Comment by dktapps** [comment]
Please don't mark comments as resolved until they've been addressed by commits on the branch

File: src/block/DeadBush.php [hunk 29]
```diff
@@ -64,4 +64,12 @@ public function getFlameEncouragement() : int{
...
 		return 100;
 	}
+
+	protected function canBeSupportedBy(Block $block) : bool{
+		$id = $block->getId();
+		return $id === BlockLegacyIds::SAND
+			|| $id === BlockLegacyIds::PODZOL
+			|| $id === BlockLegacyIds::DIRT
+			|| $id === BlockLegacyIds::TERRACOTTA;
+	}
```
Comment (dktapps) [inline src/block/DeadBush.php]: This should also be moved to a separate PR

**Comment by dktapps** [comment]
Updated description:

- this does not fix vines being placeable on the side of cacti - that requires further research
- it does not fix painting interaction with signs
- it does not fix saplings on farmland (that should be done separately)

From what I can see in my testing, this looks pretty much good to go.

**Comment by ShockedPlot7560** [comment]
Thanks for updating, I hadn't noticed that it was obsolete
Once this PR is done, I will make separate placements that have not been fixed by it

File: src/block/RedstoneWire.php [hunk 30]
```diff
@@ -45,4 +57,14 @@ public function readStateFromWorld() : void{
...
 	}
+
+	public function onNearbyBlockChange() : void{
+		if(!$this->canBeSupportedBy($this->getSide(Facing::DOWN))){
+			$this->position->getWorld()->useBreakOn($this->position);
+		}
+	}
+
+	private function canBeSupportedBy(Block $block) : bool{
+		return $block->getSupportType(Facing::UP)->hasEdgeSupport();
```
Comment (dktapps) [inline src/block/RedstoneWire.php]:
```suggestion
		return $block->getSupportType(Facing::UP)->hasCenterSupport();
```

File: src/block/Slab.php [hunk 31]
```diff
@@ -125,6 +126,15 @@ protected function recalculateCollisionBoxes() : array{
 		return [AxisAlignedBB::one()->trim($this->slabType->equals(SlabType::TOP()) ? Facing::DOWN : Facing::UP, 0.5)];
 	}
 
+	public function getSupportType(int $facing) : SupportType{
+		if($this->slabType->equals(SlabType::DOUBLE())){
+			return SupportType::FULL();
+		}elseif($facing === Facing::UP && $this->slabType->equals(SlabType::TOP()) || $facing === Facing::DOWN && $this->slabType->equals(SlabType::BOTTOM())){
```
Comment (dktapps) [inline src/block/Slab.php]:
```suggestion
		}elseif(($facing === Facing::UP && $this->slabType->equals(SlabType::TOP())) || ($facing === Facing::DOWN && $this->slabType->equals(SlabType::BOTTOM()))){
```

File: src/block/SnowLayer.php [hunk 32]
```diff
@@ -80,8 +81,15 @@ protected function recalculateCollisionBoxes() : array{
...
+	public function getSupportType(int $facing) : SupportType{
+		if(!$this->canBeReplaced()){
+			return SupportType::FULL();
+		}
+		return SupportType::NONE();
+	}
+
 	private function canBeSupportedBy(Block $b) : bool{
-		return $b->isSolid() || ($b instanceof SnowLayer && $b->isSameType($this) && $b->layers === self::MAX_LAYERS);
+		return $b->getSupportType(Facing::UP)->hasCenterSupport();
```
Comment (dktapps) [inline src/block/SnowLayer.php]: This requires full support, as far as I can tell from testing
